### PR TITLE
Add IAuthorizationPolicyProvider

### PR DESCRIPTION
**Pull request by HaoK**
Allows for dynamic policy creation
(Should we rename AuthorizationOptions => AuthorizationPolicy[Provider]Options now that its moved into a different service?)

See https://github.com/aspnet/Security/issues/670

cc @davidfowl @blowdart @divega

### CODE REVIEW

**Comment by blowdart** [comment]
AuthorizationPolicyOptions sounds fine to me, but I'll need you to help me doc the changes so we can have them available when RC2 drops.

**Comment by HaoK** [comment]
Ok AuthorizationOptions => AuthorizationPolicyOptions

**Comment by PinpointTownes** [comment]
nit: `AuthorizationPolicyOptions` is fine, but not consistent with the CORS options:  https://github.com/aspnet/CORS/blob/dev/src/Microsoft.AspNetCore.Cors/CorsOptions.cs#L12 :smile:

File: src/Microsoft.AspNetCore.Authorization/IAuthorizationPolicyProvider.cs [hunk 1]
```diff
@@ -0,0 +1,18 @@
...
+    /// <summary>
+    /// A type which can provide a <see cref="AuthorizationPolicy"/> for a particular name.
+    /// </summary>
+    public interface IAuthorizationPolicyProvider
+    {
+        /// <summary>
+        /// Gets a <see cref="AuthorizationPolicy"/> from the given <paramref name="policyName"/>
+        /// </summary>
+        /// <param name="policyName"></param>
+        /// <returns></returns>
```
Comment (davidfowl) [inline src/Microsoft.AspNetCore.Authorization/IAuthorizationPolicyProvider.cs]: Should probably be async.

**Comment by davidfowl** [comment]
Another crazy question, why not support multiple?

**Comment by HaoK** [comment]
@davidfowl RE multiple providers @divega brought this up yesterday too when we reviewed this,  what would be a rationale behavior, return the first hit?  We felt multiple providers aren't really necessary since its pretty easy to extend/reuse/wrap the default implementation and use it as a fallback.

@PinpointTownes RE AuthorizationPolicyOptions, I wouldn't be surprised if the CorsOptions was named that way due to the fact following how AuthorizationOptions was named :), so there is a slight consistency issue, but I don't think that's enough to reason to not clarify the name (assuming we all agree its moar better)

File: src/Microsoft.AspNetCore.Authorization/IAuthorizationPolicyProvider.cs (hunk 1 above)
Comment (HaoK) [inline src/Microsoft.AspNetCore.Authorization/IAuthorizationPolicyProvider.cs]: Yeah I saw Cors was async, sure I will update to async

**Comment by davidfowl** [comment]
> @PinpointTownes RE AuthorizationPolicyOptions, I wouldn't be surprised if the CorsOptions was named that way due to the fact following how AuthorizationOptions was named :), so there is a slight consistency issue, but I don't think that's enough to reason to not clarify the name (assuming we all agree its moar better)

Not sure, we may want to expand the options on there one day as we add new features. It might not be policy specific. The PolicyMap might just be one of the things on that type. I wouldn't rename it.

**Comment by HaoK** [comment]
Fair enough, reverting back to AuthorizationOptions and I'll merge

**Comment by HaoK** [comment]
00ceeb026262bc848404d056416709dc306a57df

**Pull request closed without merging**